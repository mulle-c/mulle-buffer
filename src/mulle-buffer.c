//
//  mulle_buffer.c
//  mulle-buffer
//
//  Created by Nat! on 04/11/15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#include "mulle-buffer.h"

#include "include-private.h"

#include <ctype.h>


struct mulle_buffer   *mulle_buffer_create( struct mulle_allocator *allocator)
{
   struct mulle_buffer  *buffer;

   if( ! allocator)
      allocator = &mulle_default_allocator;

   buffer = mulle_allocator_malloc( allocator, sizeof( struct mulle_buffer));
   mulle_buffer_init( buffer, allocator);
   return( buffer);
}


static inline unsigned int   hex( unsigned int c)
{
   assert( c >= 0 && c <= 0xf);
   return( c >= 0xa ? c + 'a' - 0xa : c + '0');
}


void  mulle_buffer_hexdump_line( struct mulle_buffer *buffer,
                                 uint8_t *bytes,
                                 unsigned int n,
                                 size_t counter,
                                 unsigned int options)
{
   uint8_t        *memo;
   uint8_t        *s;
   unsigned int   i;
   unsigned int   value;

   memo = bytes;

   s = mulle_buffer_advance( buffer, 10);

   if( ! (options & mulle_buffer_hexdump_offset))
   {
      *s++ = (uint8_t) hex( counter >> 28 & 0xF);
      *s++ = (uint8_t) hex( counter >> 24 & 0xF);
      *s++ = (uint8_t) hex( counter >> 20 & 0xF);
      *s++ = (uint8_t) hex( counter >> 16 & 0xF);

      *s++ = (uint8_t) hex( counter >> 12 & 0xF);
      *s++ = (uint8_t) hex( counter >> 8 & 0xF);
      *s++ = (uint8_t) hex( counter >> 4 & 0xF);
      *s++ = (uint8_t) hex( counter >> 0 & 0xF);

      *s++ = ' ';
      *s++ = ' ';
   }

   if( ! (options & mulle_buffer_hexdump_hex))
   {
      s = mulle_buffer_advance( buffer, 3 * 8);
      for( i = 0; i < 8; i++)
      {
         if( i < n)
         {
            value = *bytes++;
            *s++ = (uint8_t) hex( value >> 4);
            *s++ = (uint8_t) hex( value & 0xF);
         }
         else
         {
            *s++ = ' ';
            *s++ = ' ';
         }
         *s++ = ' ';
      }

      mulle_buffer_add_byte( buffer, ' ');

      s = mulle_buffer_advance( buffer, 3 * 8);
      for( ; i < 16; i++)
      {
         if( i < n)
         {
            value = *bytes++;

            *s++ = (uint8_t) hex( value >> 4);
            *s++ = (uint8_t) hex( value & 0xF);
         }
         else
         {
            *s++ = ' ';
            *s++ = ' ';
         }
         *s++ = ' ';
      }
   }

   if( ! (options & mulle_buffer_hexdump_ascii))
   {
      mulle_buffer_add_byte( buffer, ' ');
      mulle_buffer_add_byte( buffer, '|');

      bytes = memo;

      for( i = 0; i < 16; i++)
      {
         if( i < n)
         {
            value = *bytes++;
            if( ! isprint( value))
               value = '.';
            mulle_buffer_add_byte( buffer, (uint8_t) value);
         }
         else
            mulle_buffer_add_byte( buffer, ' ');
      }

      mulle_buffer_add_byte( buffer, '|');
   }
}


void  mulle_buffer_hexdump( struct mulle_buffer *buffer,
                            uint8_t *bytes,
                            size_t length,
                            size_t counter,
                            unsigned int options)
{
   size_t   lines;
   size_t   full_lines;
   size_t   remainder;
   size_t   i;

   lines      = (length + 15) / 16;
   full_lines = length / 16;

   for( i = 0; i < full_lines; i++)
   {
      mulle_buffer_hexdump_line( buffer, bytes, 16, counter, options);
      mulle_buffer_add_byte( buffer, '\n');
      counter += 16;
      bytes   += 16;
   }

   if( i < lines)
   {
      remainder = length - full_lines * 16;
      mulle_buffer_hexdump_line( buffer, bytes, (unsigned int) remainder, counter, options);
      mulle_buffer_add_byte( buffer, '\n');
   }
}


int   mulle_flushablebuffer_done( struct mulle_flushablebuffer *buffer)
{
   int   rval;

   rval = _mulle_flushablebuffer_flush( (struct _mulle_flushablebuffer *) buffer);
   if( rval)
      return( rval);

   _mulle_buffer_done( (struct _mulle_buffer *) buffer,
                      mulle_buffer_get_allocator( (struct mulle_buffer *) buffer));
   return( 0);
}


int   mulle_flushablebuffer_destroy( struct mulle_flushablebuffer *buffer)
{
   int   rval;

   rval = _mulle_flushablebuffer_flush( (struct _mulle_flushablebuffer *) buffer);
   if( rval)
      return( rval);

   _mulle_buffer_destroy( (struct _mulle_buffer *) buffer,
                      mulle_buffer_get_allocator( (struct mulle_buffer *) buffer));
   return( 0);
}

