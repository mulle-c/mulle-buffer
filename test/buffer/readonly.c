#include <stdio.h>

#define mulle_buffer_assert_readable( buffer) \
    do { \
        if (((buffer)->_type & MULLE_BUFFER_IS_WRITEONLY)) { \
            printf("      assert failed: %s (type: %x)\n", \
                   __FUNCTION__, \
                   (buffer)->_type); \
        } else { \
            printf("      assert succeeded: %s (type: %x)\n", \
                   __FUNCTION__, \
                   (buffer)->_type); \
        }   \
    }       \
    while(0)

#define mulle_buffer_assert_writeable( buffer) \
    do { \
        if (((buffer)->_type & MULLE_BUFFER_IS_READONLY)) { \
            printf("      assert failed: %s (type: %x)\n", \
                   __FUNCTION__, \
                   (buffer)->_type); \
        } else { \
            printf("      assert succeeded: %s (type: %x)\n", \
                   __FUNCTION__, \
                   (buffer)->_type); \
        }   \
    }       \
    while(0)


#include <mulle-buffer/mulle-buffer.h>



void test_write_operations(struct mulle_buffer *buffer)
{
    char  *s;

    printf( "   write:\n");
    mulle_buffer_add_byte(buffer, 'A');
    mulle_buffer_add_string(buffer, "Test");
    mulle_buffer_add_char(buffer, 'B');
    mulle_buffer_add_uint16(buffer, 1234);
    mulle_buffer_add_uint32(buffer, 5678);
    mulle_buffer_memset(buffer, 'C', 5);
    mulle_buffer_zero_last_byte(buffer);

    // Additional write operations
    mulle_buffer_add_c_string(buffer, "Hello");
    mulle_buffer_add_c_chars(buffer, "World", 5);
    mulle_buffer_add_string_if_empty(buffer, "NotEmpty");
    mulle_buffer_add_string_if_not_empty(buffer, "Append");
    mulle_buffer_add_string_with_maxlength(buffer, "LongString", 5);
    mulle_buffer_make_string(buffer);
    mulle_buffer_remove_all(buffer);
    mulle_buffer_remove_in_range(buffer, mulle_range_make(0, 5));
    mulle_buffer_set_length(buffer, 10);
    mulle_buffer_advance(buffer, 5);
    mulle_buffer_add_buffer(buffer, NULL);
    mulle_buffer_add_buffer_range(buffer, NULL, mulle_range_make(0, 5));

    mulle_buffer_get_string(buffer);

    // write a zero if needed
    s = mulle_buffer_extract_string(buffer);
    mulle_free( s);
}

void test_read_operations(struct mulle_buffer *buffer)
{
    char tmp[ 10];

    printf( "   read:\n");

    mulle_buffer_get_last_byte(buffer);
    mulle_buffer_get_byte(buffer, 0);

    // Additional read operations
    mulle_buffer_get_bytes(buffer);
    mulle_buffer_next_byte(buffer);
    mulle_buffer_next_character(buffer);
    mulle_buffer_peek_byte(buffer);
    mulle_buffer_seek_byte(buffer, 'A');
    mulle_buffer_next_bytes(buffer, tmp, 10);
    mulle_buffer_reference_bytes(buffer, 5);
}

void test_harmless_operations(struct mulle_buffer *buffer)
{
    printf( "   harmless:\n");
    mulle_buffer_is_empty(buffer);
    mulle_buffer_get_length(buffer);
    mulle_buffer_get_bytes(buffer);

    // Additional harmless operations
    mulle_buffer_get_capacity(buffer);
    mulle_buffer_get_allocator(buffer);
    mulle_buffer_get_seek(buffer);
    mulle_buffer_set_seek(buffer, SEEK_SET, 0);
    mulle_buffer_get_staticlength(buffer);
}


void  test( int readonly)
{
    struct mulle_buffer buffer;
    mulle_buffer_init_default(&buffer);
    if( readonly)
    {
       mulle_buffer_set_readonly(&buffer);
       printf( "READONLY:\n");
    }
    else
    {
       mulle_buffer_set_writeonly(&buffer);
       printf( "WRITEONLY:\n");
    }

    test_write_operations(&buffer);
    test_read_operations(&buffer);
    test_harmless_operations(&buffer);
    
    mulle_buffer_done(&buffer);
}


int  main( void)
{
   test( 0);
   test( 1);
   return( 0);
}