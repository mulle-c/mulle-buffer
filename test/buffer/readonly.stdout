WRITEONLY:
   write:
      assert succeeded: mulle_buffer_add_byte (type: 80)
      assert succeeded: mulle_buffer_add_string (type: 80)
      assert succeeded: mulle_buffer_add_char (type: 80)
      assert succeeded: mulle_buffer_add_uint16 (type: 80)
      assert succeeded: mulle_buffer_add_uint32 (type: 80)
      assert succeeded: mulle_buffer_memset (type: 80)
      assert succeeded: mulle_buffer_zero_last_byte (type: 80)
      assert succeeded: mulle_buffer_add_c_string (type: 80)
      assert succeeded: mulle_buffer_add_c_chars (type: 80)
      assert succeeded: mulle_buffer_add_string_if_empty (type: 80)
      assert succeeded: mulle_buffer_add_string_if_not_empty (type: 80)
      assert succeeded: mulle_buffer_add_string_with_maxlength (type: 80)
      assert succeeded: mulle_buffer_make_string (type: 80)
      assert succeeded: mulle_buffer_remove_all (type: 80)
      assert succeeded: mulle_buffer_remove_in_range (type: 80)
      assert succeeded: mulle_buffer_set_length (type: 80)
      assert succeeded: mulle_buffer_get_string (type: 80)
      assert succeeded: mulle_buffer_extract_string (type: 80)
   read:
      assert failed: mulle_buffer_get_last_byte (type: 80)
      assert failed: mulle_buffer_get_byte (type: 80)
      assert failed: mulle_buffer_next_byte (type: 80)
      assert failed: mulle_buffer_next_character (type: 80)
      assert failed: mulle_buffer_peek_byte (type: 80)
      assert failed: mulle_buffer_seek_byte (type: 80)
      assert failed: mulle_buffer_next_bytes (type: 80)
      assert failed: mulle_buffer_reference_bytes (type: 80)
   harmless:
READONLY:
   write:
      assert failed: mulle_buffer_add_byte (type: 40)
      assert failed: mulle_buffer_add_string (type: 40)
      assert failed: mulle_buffer_add_char (type: 40)
      assert failed: mulle_buffer_add_uint16 (type: 40)
      assert failed: mulle_buffer_add_uint32 (type: 40)
      assert failed: mulle_buffer_memset (type: 40)
      assert failed: mulle_buffer_zero_last_byte (type: 40)
      assert failed: mulle_buffer_add_c_string (type: 40)
      assert failed: mulle_buffer_add_c_chars (type: 40)
      assert failed: mulle_buffer_add_string_if_empty (type: 40)
      assert failed: mulle_buffer_add_string_if_not_empty (type: 40)
      assert failed: mulle_buffer_add_string_with_maxlength (type: 40)
      assert failed: mulle_buffer_make_string (type: 40)
      assert failed: mulle_buffer_remove_all (type: 40)
      assert failed: mulle_buffer_remove_in_range (type: 40)
      assert failed: mulle_buffer_set_length (type: 40)
      assert failed: mulle_buffer_get_string (type: 40)
      assert failed: mulle_buffer_extract_string (type: 40)
   read:
      assert succeeded: mulle_buffer_get_last_byte (type: 40)
      assert succeeded: mulle_buffer_get_byte (type: 40)
      assert succeeded: mulle_buffer_next_byte (type: 40)
      assert succeeded: mulle_buffer_next_character (type: 40)
      assert succeeded: mulle_buffer_peek_byte (type: 40)
      assert succeeded: mulle_buffer_seek_byte (type: 40)
      assert succeeded: mulle_buffer_next_bytes (type: 40)
      assert succeeded: mulle_buffer_reference_bytes (type: 40)
   harmless:
